http-client/Network/HTTP/Client/Connection.hs

```diff
@@ -184,10 +189,31 @@ openSocket tweakSocket addr =
             NS.connect sock (NS.addrAddress addr)
             return sock)
 
+-- Pick up an IP using an approximation of the happy-eyeballs algorithm:
+-- https://datatracker.ietf.org/doc/html/rfc8305
+-- 
 firstSuccessful :: [NS.AddrInfo] -> (NS.AddrInfo -> IO a) -> IO a
-firstSuccessful []     _  = error "getAddrInfo returned empty list"
-firstSuccessful (a:as) cb =
-    cb a `E.catch` \(e :: E.IOException) ->
-        case as of
-            [] -> E.throwIO e
-            _  -> firstSuccessful as cb
+firstSuccessful []        _  = error "getAddrInfo returned empty list"
+firstSuccessful addresses cb = do
+    result <- newEmptyMVar
+    either E.throwIO pure =<<
+        withAsync (tryAddresses result)
+            (\_ -> takeMVar result)
+  where
+    -- https://datatracker.ietf.org/doc/html/rfc8305#section-5
+    connectionAttemptDelay = 250 * 1000
+
+    tryAddresses result = do
+        lastException <- take 1 . filter isLeft <$> go addresses []
+        for_ lastException $ tryPutMVar result
+      where
+        go []             asyncs = mapM wait asyncs
+        go (addr : addrs) asyncs =
+            withAsync (E.evaluate =<< tryAddress addr) $ \a -> do
+                unless (null addrs) $ threadDelay connectionAttemptDelay
```

Comment:
This seems to say that only the _final_ address won't have a delay.

Reply:
The intention here is "if there're no more addresses to consider, don't have a delay". So if, lets say. there're 3 addresses it winds up looking like this

```
withAsync (E.evaluate =<< tryAddress addr0) $ \a0 -> do
  threadDelay connectionAttemptDelay
  withAsync (E.evaluate =<< tryAddress addr1) $ \a1 -> do
    threadDelay connectionAttemptDelay
    withAsync (E.evaluate =<< tryAddress addr2) $ \a2 ->
       -- no delay here
       mapM wait [a2, a1, a0]  
```

Unless one of `[a2, a1, a0]` fills in the `result` and then this whole stack gets cancelled by the top-level `withAsync(tryAddresses result) ... `.

Reply:
This is all heading back in the "too complicated" direction I originally was concerned about. I'm not convinced the code is behaving correctly, especially with the exception handling logic.